[FIX] account_asset_asset: compute_depreciation_board

nowheresly opened the pull request:
When the date of the last depreciation line marked as init entry does not match the asset start date, the compute_depreciation_board method fails to a linear depreciation. This fix ensures the depreciation remains linear in such case.
For example, let's consider the following asset:
![selection_001](https://cloud.githubusercontent.com/assets/916109/8398238/b6a0036a-1de6-11e5-9a86-0fd3de61f4cd.png)

Some of the depreciation of this asset has already been accounted using another accounting software.
So we have a depreciation line marked as init_entry=True with a depreciation amount corresponding to:
- the depreciation from all the previous (closed) periods;
- the depreciation for the current (still opened) period.
  (see picture below)
  ![selection_002](https://cloud.githubusercontent.com/assets/916109/8398262/67cd53c2-1de7-11e5-834e-5b4504554adf.png)

In such case, when we hit the compute button, we have the following behaviour:
![selection_003](https://cloud.githubusercontent.com/assets/916109/8398266/72dfeaa4-1de7-11e5-8a65-2912856f5489.png)

As you can see, the last depreciation line of the current period is wrong. The method `compute_depreciation_board` from `account_asset_asset` has erroneously taken into account the depreciation coming from the previous period.

This fix just removes the depreciation part coming from the older periods to ensure a linear depreciation.

nowheresly committed:
[IMP] account_asset_asset: add unit test for compute_depreciation_board when there are
several depreciation lines with init_entry=True.

laetitia-gangloff commented:
Hello,

I create a new asset (TEST-LGA) with the following configuration :
- depreciation on 5 years with month period
- begin at 22nd november 2013
- for an amount of 120 000
- and a depreciation line for december 2014 of 28 000

When I hit the "compute" button I have the following result :

![image](https://cloud.githubusercontent.com/assets/7602419/8411927/d3276226-1e86-11e5-835a-ad219c04ad0e.png)

Is  it correct to get 30 000 in the last period of the current fiscal year ?

sbidoul commented:
@laetitia-gangloff which depreciation method (linear, ...)?, pro-rata temporis or not?

sbidoul commented:
@laetitia-gangloff I think your problem is different and possibly 7.0 specific (I can't reproduce on 8.0).

@Nowheresly there seem to be something wrong indeed. Not 100% sure about your fix though. Must dig further.

luc-demeyer commented:
@Nowheresly 
I programmed this calculation (a long time ago) and tested the init flag in several use cases but I never tested the case with a fiscal year without depreciation. The idea is that you should be able to generate the correct legal reports, hence I would expect an init entry at the end of 2014, and a second init entry in 2015 (which doesn't mean that your use case shouldn't work).
I tried to reproduce your case on my V8 test system but the calculate gave in my test another result. What are the Fiscal Year start/stop dates for 2014 and 2015 ?

laetitia-gangloff commented:
@sbidoul linear, no pro-rata temporis

![image](https://cloud.githubusercontent.com/assets/7602419/8426490/6efbee86-1f10-11e5-854d-3ba6a7ac9bbb.png)

It was on this runbot, so I didn't think it is 7.0 specific.

sbidoul commented:
@laetitia-gangloff I understand now. So your scenario in https://github.com/OCA/account-financial-tools/pull/249#issuecomment-116738440 reveals a regression introduced by this PR.

nowheresly commented:
@sbidoul @luc-demeyer @laetitia-gangloff  thanks for taking time to review my work and for your feedback !

@luc-demeyer : fiscal year starts january 1st and ends december 31st.
You're right: I splitted the depreciate line init entry in two different depreciation lines:
- one for all previous fiscal periods with a line_date=31/12/<previous year>
- one for the current fiscal period.

After the splitting the calculated depreciation board is correct !
I modified my PR accordingly:
- I added a check to avoid negative depreciation amount
- I added 2 unit tests to describe the issue

luc-demeyer commented:
@Nowheresly 
You should remove check on negative depreciation amount. These are perfectly valid. 
E.g. you made a mistake and depreciated too much -> should be compensated by negative amount. Another example is changing depreciation duration (if the duration becomes longer, there will be probably be negative depreciation amounts).

Thanks for adding unit tests. Seen the large number of use cases and the limited amount of tests, we still have a way to go before we have a complete set of tests.

nowheresly commented:
@luc-demeyer thanks for the explanations. 
Adding a unit test was very easing using those already written :)
So I removed the check on negative depreciation amount.
Not sure about what we could do to prevent wrong calculations. 
At least, I added a comment trying to explain how to manage this specific case...

laetitia-gangloff commented on account_asset_management/account_asset.py:
```diff
@@ -605,6 +605,20 @@ def _get_depreciation_entry_name(self, cr, uid, asset, seq, context=None):
+
+        CAUTION: make sure the last depreciation line with init_entry=True
+        contains only depreciated amount for its respective fiscal year or this
+        method may compute wrong depreciation lines.
```
This is also possible to complete the help on the field "Asset Start Date" in the view ?

luc-demeyer commented:
@Nowheresly @laetitia-gangloff 
Can you have a look at https://github.com/OCA/account-financial-tools/pull/296 to see if this change fixes the issues reported in this PR.
If so than this PR becomes obsolute except for the unit test added in this PR.
But looking to it, this unit test should be corrected : 
- negative values can be perfectly valid, hence we shouldn't test on that
- I tested the use case of the unit test added in the PR 296 version and verified with XLS calculation.
  This is what the unit test should check:
  Asset Value 1392,00         
  
  ```
  depr    amount  remains     
  ```
  
  2014-12-31  0,00    87,00   1305,00     init_entry
  2015-05-31  87,00   144,70  1160,30     init_entry
  2015-12-31  231,70  203,30  957,00  
  2016-12-31  435,00  348,00  609,00  
  2017-12-31  783,00  348,00  261,00  
  2018-12-31  1131,00 261,00  0,00

nowheresly commented:
@laetitia-gangloff , @luc-demeyer 
Sorry for my late reply. I will check your PR fix mine.

adrienpeiffer commented:
Fixed on #296
